sl1m/qp.py

```diff
@@ -13,55 +27,290 @@ def quadprog_solve_qp(P, q, G=None, h=None, C=None, d=None, verbose = False):
     qp_a = -q
     if C is not None:
         if G is not None:
-                qp_C = -vstack([C, G]).T
-                qp_b = -hstack([d, h])   
+            qp_C = -vstack([C, G]).T
+            qp_b = -hstack([d, h])
         else:
-                qp_C = -C.transpose()
-                qp_b = -d 
+            qp_C = -C.transpose()
+            qp_b = -d
         meq = C.shape[0]
-    else:  # no equality constraint 
+    else:  # no equality constraint
         qp_C = -G.T
         qp_b = -h
-        meq = 0 
+        meq = 0
     res = quadprog.solve_qp(qp_G, qp_a, qp_C, qp_b, meq)
     if verbose:
-            return res
+        return res
     #~ print 'qp status ', res
     return res[0]
 
 
-#min ||Ax-b||**2 
+#min ||Ax-b||**2
 #subject to  G x <= h
 #subject to  C x  = d
 def solve_least_square(A,b,G=None, h=None, C=None, d=None):
-        P = dot(A.T, A)
-        #~ q = 2*dot(b, A).reshape(b.shape[0])
-        q = -dot(A.T, b).reshape(b.shape[0])
-        #~ q = 2*dot(b, A) 
-        return quadprog_solve_qp(P, q, G, h, C, d)
+    P = dot(A.T, A)
+    #~ q = 2*dot(b, A).reshape(b.shape[0])
+    q = -dot(A.T, b).reshape(b.shape[0])
+    #~ q = 2*dot(b, A)
+    return solve_qp_quadprog(P, q, G, h, C, d)
+
+#min x' C x + c x
+#subject to  A x <= b
+#subject to  E x  = e
+def gurobi_solve_qp(C, c, A=None, b=None, E=None, e=None):
+    if not GUROBI_OK:
+        print("Requires gurobi")
+        raise ImportError
+    model = grb.Model("lp")
+
+    #add continuous variables
+    cVars = []
+    for el in (c):
+        cVars.append(model.addVar(vtype=grb.GRB.CONTINUOUS,  lb = -grb.GRB.INFINITY, ub = grb.GRB.INFINITY))
+
+    # Update model to integrate new variables
+    model.update()
+    x = array(model.getVars(), copy=False)
+
+    # equality constraints
+    if E.shape[0] > 0:
+        for i in range(E.shape[0]):
+            idx = [j for j, el in enumerate(E[i].tolist()) if el != 0.]
+            variables = x[idx]
+            coeff = E[i,idx]
+            expr = grb.LinExpr(coeff, variables)
+            model.addConstr(expr, grb.GRB.EQUAL, e[i])
+    model.update()
+
+    # inequality constraints
+    if A.shape[0] > 0:
+        for i in range(A.shape[0]):
+            idx = [j for j, el in enumerate(A[i].tolist()) if el != 0.]
+            variables = x[idx]
+            coeff = A[i,idx]
+            expr = grb.LinExpr(coeff, variables)
+            model.addConstr(expr, grb.GRB.LESS_EQUAL, b[i])
+    model.update()
+
+    # set objective
+    obj = grb.QuadExpr()
+    for i in range(C.shape[0]):
+        for j in range(C.shape[1]):
+            if C[i][j] != 0:
+                obj += C[i][j]*cVars[i]*cVars[j]
```

Comment:
```suggestion
                obj += C[i][j] * cVars[i] * cVars[j]
```

Reply:
There are many places where some similar spaces are missing. But don't worry about those, a run of an automatic formatter like yapf will fix this. And I can handle it if you want.